### DISCUSSION TITLE
Rework type guards in Radashi

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#46] [Ideas]
There are libraries like `@sindresorhus/is` out there that are more widely used, which are therefore more battle-tested/robust and (perhaps most importantly) more comprehensive.

I suggest copying my code from https://github.com/alloc/is/blob/master/src/is.ts (which is written to be tree-shakeable and not rely on NodeJS types anywhere). Then we can update its implementation to align with `@sindresorhus/is` to achieve the same robustness without the baggage.

### Related issues
- https://github.com/sodiray/radash/issues/405

**Comment by nnmrts** [comment]
For return types radashi could also profit from https://github.com/sindresorhus/type-fest.

**Reply by aleclarson** [reply]
Appreciate the suggestion! 👍  We might already have types from there even!

**Comment by baxyz** [comment]
Just to keep trace:

| methods                          | radashi | alloc/is | @sindresorhus/is |
| -------------------------------- | ------- | -------- | ---------------- |
| (global) is / detect             |         |          | ☑️               |
| (global) assert                  |         |          | ☑️               |
| isAll                            |         |          | ✅               |
| isAny                            |         |          | ✅               |
| isArray                          | ✅      | ✅       | ✅               |
| isArrayBuffer                    |         |          | ✅               |
| isArrayLike                      |         |          | ✅               |
| isAsyncFunction                  |         | ✅       | ✅               |
| isAsyncGenerator                 |         |          | ✅               |
| isAsyncGeneratorFunction         |         |          | ✅               |
| isAsyncIterable                  |         | ✅       | ✅               |
| isBigint                         |         | ✅       | ✅               |
| isBigInt64Array                  |         |          | ✅               |
| isBigUint64Array                 |         |          | ✅               |
| isBlob                           |         |          | ✅               |
| isBoolean                        | ✅      | ✅       | ✅               |
| isBoundFunction                  |         |          | ✅               |
| isBuffer                         |         |          | ✅               |
| isClass                          | ✅      | ✅       | ✅               |
| isDataView                       |         |          | ✅               |
| isDate                           | ✅      | ✅       | ✅               |
| isDefined                        |         | ✅       |                  |
| isDirectInstanceOf               |         |          | ✅               |
| isEmpty                          | ✅      |          |                  |
| isEmptyArray                     |         |          | ✅               |
| isEmptyMap                       |         |          | ✅               |
| isEmptyObject                    |         | ✅       | ✅               |
| isEmptySet                       |         |          | ✅               |
| isEmptyString                    |         |          | ✅               |
| isEmptyStringOrWhitespace        |         |          | ✅               |
| isEnumCase                       |         |          | ✅               |
| isEqual                          | ✅      |          |                  |
| isError                          | ✅      | ✅       | ✅               |
| isEvenInteger                    |         |          | ✅               |
| isFalsy                          |         |          | ✅               |
| isFloat                          | ✅      |          |                  |
| isFloat32Array                   |         |          | ✅               |
| isFloat64Array                   |         |          | ✅               |
| isFormData                       |         |          | ✅               |
| isFunction                       | ✅      | ✅       | ✅               |
| isGenerator                      |         | ✅       | ✅               |
| isGeneratorFunction              |         | ✅       | ✅               |
| isHtmlElement                    |         |          | ✅               |
| isInfinite                       |         | ✅       | ✅               |
| isInRange                        |         |          | ✅               |
| isInt / isInteger                | ✅      | ✅       | ✅               |
| isInt16Array                     |         |          | ✅               |
| isInt32Array                     |         |          | ✅               |
| isInt8Array                      |         |          | ✅               |
| isIntString                      | ✅      |          |                  |
| isIterable                       | ✅      | ✅       | ✅               |
| isMap                            | ✅      | ✅       | ✅               |
| isNaN                            |         | ✅       | ✅               |
| isNativePromise                  |         |          | ✅               |
| isNegativeNumber                 |         |          | ✅               |
| isNodeStream                     |         |          | ✅               |
| isNonEmptyArray                  |         |          | ✅               |
| isNonEmptyMap                    |         |          | ✅               |
| isNonEmptyObject                 |         |          | ✅               |
| isNonEmptySet                    |         |          | ✅               |
| isNonEmptyString                 |         |          | ✅               |
| isNonEmptyStringAndNotWhitespace |         |          | ✅               |
| isNull                           |         | ✅       | ✅               |
| isNullish / isNullOrUndefined    | ✅      |          | ✅               |
| isNumber                         | ✅      | ✅       | ✅               |
| isNumericString                  |         |          | ✅               |
| isObject                         | ✅      |          | ✅               |
| isObjectOrFunction               |         | ✅       |                  |
| isObservable                     |         |          | ✅               |
| isOddInteger                     |         |          | ✅               |
| isPlainObject                    | ✅      | ✅       | ✅               |
| isPositiveNumber                 |         |          | ✅               |
| isPrimitive                      | ✅      |          | ✅               |
| isPromise                        | ✅      | ✅       | ✅               |
| isPromiseLike                    |         | ✅       |                  |
| isPropertyKey                    |         |          | ✅               |
| isRegExp                         | ✅      | ✅       | ✅               |
| isResult                         | ✅      |          |                  |
| isResultErr                      | ✅      |          |                  |
| isResultOk                       | ✅      |          |                  |
| isSafeInteger                    |         | ✅       | ✅               |
| isSet                            | ✅      | ✅       | ✅               |
| isSharedArrayBuffer              |         |          | ✅               |
| isString                         | ✅      | ✅       | ✅               |
| isSymbol                         | ✅      | ✅       | ✅               |
| isTagged                         | ✅      |          |                  |
| isTruthy                         |         |          | ✅               |
| isTupleLike                      |         |          | ✅               |
| isTypedArray                     |         |          | ✅               |
| isUint16Array                    |         |          | ✅               |
| isUint32Array                    |         |          | ✅               |
| isUint8Array                     |         |          | ✅               |
| isUint8ClampedArray              |         |          | ✅               |
| isUndefined                      | ✅      | ✅       | ✅               |
| isUrlInstance                    |         |          | ✅               |
| isUrlSearchParams                |         |          | ✅               |
| isUrlString                      |         |          | ✅               |
| isValidDate                      |         |          | ✅               |
| isValidLength                    |         |          | ✅               |
| isWeakMap                        | ✅      | ✅       | ✅               |
| isWeakRef                        |         |          | ✅               |
| isWeakSet                        | ✅      | ✅       | ✅               |
| isWhitespaceString               |         |          | ✅               |
|                                  | 30      | 32       | 90 + 2           |
